warehouse/query-core/src/main/java/datawave/query/jexl/nodes/QueryPropertyMarker.java

```diff
@@ -161,6 +165,66 @@ protected void setupSource(JexlNode source) {
         this.jjtAddChild(refExpNode1, 0);
     }
     
+    /**
+     * Determines if this source node is already marked with the provided marker class
+     *
+     * @param source
+     *            a JexlNode
+     * @param markerClass
+     *            an instance of a {@link QueryPropertyMarker}
+     * @return true if this node is already marked with the provided marker class
+     */
+    public static boolean isSourceAlreadyMarked(JexlNode source, Class<? extends QueryPropertyMarker> markerClass) {
+        return QueryPropertyMarker.findInstance(source).isType(markerClass);
+    }
+    
+    /**
+     * Determine if one the source node's ancestors is already marked. For example, is this subtree already marked as delayed?
+     *
+     *
+     * @param node
+     *            a JexlNode
+     * @return true
+     */
+    public static boolean isSubTreeAlreadyMarked(JexlNode node, Class<? extends QueryPropertyMarker> markerClass) {
+        // marker nodes are by definition the intersection of a marker and a source
+        if (node instanceof ASTAndNode && node.jjtGetNumChildren() == 2) {
+            if (QueryPropertyMarker.findInstance(node).isType(markerClass)) {
+                return true;
+            }
+        }
+        
+        if (node.jjtGetParent() == null) {
+            return false;
+        } else {
+            return isSubTreeAlreadyMarked(node.jjtGetParent(), markerClass);
+        }
+    }
+    
+    /**
+     * Unwrap a delayed predicate, fully. Intended to handle the odd edge case when multiple delayed predicate markers are applied to the same node
+     *
+     * @param node
+     *            an arbitrary jexl node
+     * @return the source node, or the original node if this node is not delayed
+     */
+    public static JexlNode unwrapFully(JexlNode node) {
```

Comment:
I would like to see this method moved to ASTDelayedPredicate to further imply that this is an operation that specifically revolves around delayed predicate markers. Additionally, this can be simplified to
```java
public static JexlNode unwrapFully(JexlNode node) {
    QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
    if (instance.isType(ASTDelayedPredicate.class)) {
        return unwrapFully(instance.getSource());
    }
    return node;
}
```

Reply:
I'm looking at refactoring this to `unwrapFully(node, clazz)` so the caller can specify a marker class. I think that case would address your comment?

Reply:
Yes it would. I'm fine with leaving this in QueryPropertyMarker in that case.